### ISSUE TITLE
[BUG] PyTorch Data loader do not reinitialize properly when iterating over it a second time

### ISSUE THREAD

**Issue by gabrielspmoreira** [NVIDIA/NVTabular#513]
**Describe the bug**
When you iterate over a dataloader (shuffle = False) and then start iterating again over it (for example, for another training epoch), the expected behavior would be that the dataloader internal cursor is reset, and you get the same batches again, like the regular PyTorch [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader).
Therefore, when the dataloader is not fully iterated, when it is iterated again the first batch is the last batch of the dataset (which has usually less rows than the batch size, for being the last of the dataset).

**Steps/Code to reproduce bug**

```python
from nvtabular.loader.torch import TorchAsyncItr 
from nvtabular import Dataset

data_loader_config = {
                "cats": categ_features,
                "conts": continuous_features,
                "labels": [],
                "devices": list(range(training_args.n_gpu)),
            }

#Assume the train set has 107 rows
train_set = Dataset(train_data_path, engine="parquet", part_mem_fraction=data_args.nvt_part_mem_fraction)
BATCH_SIZE = 10
train_loader = TorchAsyncItr(dataset=train_set, 
                                            batch_size=BATCH_SIZE, 
                                            shuffle=shuffle_train_dataloader, **data_loader_config)

print('*** First iteration ***')
for step, inputs for enumerate(train_loader):
    if step <= 2:
        cats. conts, labels = inputs
        print('shape', conts[1]['a_feature_name'][1].shape[0])
    else:
        break


print('*** Second iteration ***')
for step, inputs for enumerate(train_loader):
    if step <= 2:
        cats. conts, labels = inputs
        print('shape', conts[1]['a_feature_name'][1].shape[0])
    else:
        break
```

The output will be something like

```
*** First iteration ***
shape 10
shape 10

*** Second iteration ***
shape 7
shape 10
```
In this case, if we assume the dataset has length 107 and a batch size of 10, the first batch of the "Second iteration" will be the last one of the dataset (incomplete, with only 7 examples) and the second batch will be equal to the first batch of the "First iteration"

**Expected behavior**
Subsequent iterations on the same dataloader should return the same results, and the "incomplete" batch should be only the last one.

**Environment details (please complete the following information):**
NVTabular 0.3 installed with conda
 
**Additional comments**

Related Issue: #470 - [FEA] Drop_last batch option for PyTorch and TF Dataloaders
